src/Bladeburner/Actions/ActionIdentifier.ts

```diff
@@ -0,0 +1,66 @@
+import type { ActionIdentifier } from "../Types";
+import { BladeburnerActionType } from "@enums";
+import { assertLoadingType } from "../../utils/TypeAssertion";
+import { getEnumHelper } from "../../utils/EnumHelper";
+
+/** Loads an action identifier
+ * This is used for loading ActionIdentifier class objects from pre-2.6.1
+ * Should load both the old format and the new format */
+export function resolveActionIdentifier(identifier: unknown): ActionIdentifier | null {
+  if (!identifier || typeof identifier !== "object") return null;
+  assertLoadingType<ActionIdentifier>(identifier);
+  for (const matcher of TypeToActionMatchers) {
+    const id = matcher(identifier.name);
+    if (id) {
+      return id;
+    }
+  }
+
+  return null;
+}
+
+/** These shorthands match those documented in the BB Terminal Help */
+export const TerminalShorthands = {
+  [BladeburnerActionType.Contract]: <readonly string[]>["contract", "contracts", "contr"],
+  [BladeburnerActionType.Operation]: <readonly string[]>["operation", "operations", "op", "ops"],
+  [BladeburnerActionType.BlackOp]: <readonly string[]>[
+    "blackoperation",
+    "black operation",
+    "black operations",
+    "black op",
+    "black ops",
+    "blackop",
+    "blackops",
+  ],
+  [BladeburnerActionType.General]: <readonly string[]>["general", "general action", "gen"],
+} as const;
+
+export function autoCompleteTypeShorthand(typeShorthand: string, name: string): ActionIdentifier | null {
+  const matchedType = typeShorthand.toLowerCase().trim();
+  let id = resolveActionIdentifier({ name });
+
+  if (id && !TerminalShorthands[id.type].includes(matchedType)) {
+    id = null;
+  }
+
+  return id;
+}
+
+const TypeToActionMatchers = [
```

Comment:
Rather than this being a bunch of closures with redundant info, I think it would be better to have it be a simple array of pairs; The name of the enum and the ActionType are the only things that are changing here. The rest of the code can be pulled to resolveActionIdentifier.

You can look at the code I recently posted in #suggestions for guidance if you are having type issues.

Reply:
the lambda list was also something I was worrying about, maybe something like this could be a middle ground:
```ts
function getTypedKeys<T extends object>(obj: T): Array<keyof T> {
  return Object.keys(obj) as Array<keyof T>;
}

const ActionEnums = {
  General: BladeburnerGeneralActionName,
  Contract: BladeburnerContractName,
  Operation: BladeburnerOperationName,
  BlackOp: BladeburnerBlackOpName,
};

const getActionEnum = (key: keyof typeof BladeburnerActionType) => ActionEnums[key];

const inActionEnum = (key: keyof typeof BladeburnerActionType, name: string) =>
  Object.values(getActionEnum(key)).includes(name);

const matchTypeToAction = (key: keyof typeof BladeburnerActionType, name: string): ActionIdentifier | null =>
  inActionEnum(key, name) ? ({ type: BladeburnerActionType[key], name: name } as ActionIdentifier) : null;
  // I know 'as' is circumventing ts here, but you should be able to see that this typechecks.

export const resolveActionIdentifier = (identifier: string): ActionIdentifier | null => {
  const keys = getTypedKeys(BladeburnerActionType);

  for (const key of keys) {
    const matchedAction = matchTypeToAction(key, identifier);
    if (matchedAction) {
      return matchedAction;
    }
  }

  return null; // Return null if no match is found

  // I'd prefer reduce ;D but I know that's not everyone's cup of tea
  return getTypedKeys(BladeburnerActionType).reduce<ActionIdentifier | null>((acc, key) => {
    return acc ?? matchTypeToAction(key, identifier);
  }, null);
};
```

Reply:
> Rather than this being a bunch of closures with redundant info, I think it would be better to have it be a simple array of pairs; The name of the enum and the ActionType are the only things that are changing here. The rest of the code can be pulled to resolveActionIdentifier.
> 
> You can look at the code I recently posted in #suggestions for guidance if you are having type issues.

Can you provide an example?

Reply:
Nevermind! I did some root cause analysis and it looks like we're solving the wrong problem. I'll add type discriminator guards to where the types are used, rather than the factory.

Reply:
```js
updateSkillLevels(this: Person): void {
  const skillsAndMults: readonly (readonly [keyof Skills & keyof Multipliers, keyof BitNodeMultipliers])[] = [
    ["hacking", "HackingLevelMultiplier"],
    ["strength", "StrengthLevelMultiplier"],
    ["defense", "DefenseLevelMultiplier"],
    ["dexterity", "DexterityLevelMultiplier"],
    ["agility", "AgilityLevelMultiplier"],
    ["charisma", "CharismaLevelMultiplier"],
  ] as const;
  for (const [skill, bnMult] of skillsAndMults) {
    this.skills[skill] = Math.max(
      1,
      Math.floor(this.calculateSkill(this.exp[skill], this.mults[skill] * currentNodeMults[bnMult])),
    );
  }
  
  const ratio: number = Math.min(this.hp.current / this.hp.max, 1);
  this.hp.max = Math.floor(10 + this.skills.defense / 10);
  this.hp.current = Math.round(this.hp.max * ratio);
}
```

Here's the code I was referring to. It's solving a slightly different problem, but it might give you some ideas.

Reply:
Ah I see. My case is a bit different. We have 3 types: The Enum, the name and the type.

The type and name form a discriminated union to cover the 4 types as the names are all type-specific.

The reason I put it in a closure previously is to maintain the link between all three. Using your example only 2 of the inferences are maintained, and one has to be cast. I'm hesitant to go to a solution that involves more complex typing while losing precision. I'm leaning to simply deduplicate the closures. How will this affect your stance?

Reply:
The root cause of the issue is that a discriminated union type is being used to extract runtime information about the type and put the type back into the specific context. The union doesn't share a parent class, so it's not subtype polymorphism, nor is the parent type desired. The union is only used together or specifically. I'll address that in the next update.

The issue is that when the union is being iterated over, the subtype properties (type and name) are being split into a discriminated union as well and lose the pairing. This makes it impossible to put them back together without a type guard that checks both or an unsafe cast. I've created the closures to act as a type guard over the whole. But I agree they have too much logic.

Reply:
I'd still love to know why
`inActionEnum(key, name) ? ({ type: BladeburnerActionType[key], name: name } as ActionIdentifier) : null;`
is bad. ActionIdentifier acts as a glue between two enums: BladeburnerActionType and enums for each of it's values. So, in every other 'normal' programming language that would just a pair of integers (BladeburnerActionType, BladeburnerActionSubType). In ts string enums we now have to check for validity of those pairs at runtime. The user inputs those values via the console - they cannot be typechecked at compile time.
That only valid pairs can be constructued is completly obvious in:
`inActionEnum(key, name) ? ({ type: BladeburnerActionType[key], name: name } as ActionIdentifier) : null;`
It's saying: Is `name` in the BladeburnerActionType sub-enum `key`? If so, form the **valid** pair. 

To prevent unnoticed strange changes to the ActionIdentifier type I'd suggest moving this check closely to where the type is defined.
----
If I was to completly rewrite bb, I'd drop the name enum(s) altogether. It is, in it's own, a rip out of the action data files (like in bladeburner/data/contracts.ts): `[BladeburnerContractName.Tracking]: new Contract({
      name: BladeburnerContractName.Tracking`, which should simply be part of a list [new Contract{name : "Tracking"}], and all other needed datastructures, like a list of all Contrac names, should be derived from that.

Reply:
That's a great question. AI would be fine if it was a supertype or at least common denomitator. That's what it _looks like_ at first glance.

But it's not. It is a discriminated union (or tagged type), so there is no "all of them" state. It's specifically _excatly one of_ the four, not a denominator that can hold _either one of_
That detail is crucial because it makes the type erasure when going inside a loop split the properties (name and type) also a union, rather than a denominator.

This problem just keeps getting worse the more hard-casts and separation we keep putting in to circumvent it because it requires switch statements on both ends.

Instead, I'm making the problem go away without having to keep carrying around. Ultimately, I think we have the same direction in mind but your root cause analysis is incomplete.

Reply:
New update is up, made it simpler.